src/main/kotlin/no/nav/tilleggsstonader/sak/vilkår/stønadsperiode/StønadsperiodeService.kt

```diff
@@ -1,38 +1,90 @@
-        stønadsperiodeRepository.deleteAll() // TODO skal vi finne og oppdatere de som har blitt oppdatert/slettet
-        return stønadsperiodeRepository.insertAll(
-            stønadsperioder.map {
-                Stønadsperiode(
-                    id = it.id ?: UUID.randomUUID(),
-                    behandlingId = behandlingId,
+
+        val tidligereStønadsperioder = stønadsperiodeRepository.findAllByBehandlingId(behandlingId)
+
+        slettStønadsperioder(tidligereStønadsperioder, stønadsperioder)
+        val oppdaterteStønadsperioder = oppdaterStønadsperioder(tidligereStønadsperioder, stønadsperioder)
+        val nyeStønadsperioder = nyeStønadsperioder(behandlingId, stønadsperioder, tidligereStønadsperioder)
+        return (nyeStønadsperioder + oppdaterteStønadsperioder).tilSortertDto()
+    }
+
+    private fun slettStønadsperioder(
+        tidligereStønadsperioder: List<Stønadsperiode>,
+        stønadsperioder: List<StønadsperiodeDto>,
+    ) {
+        val gjenståendePerioder = stønadsperioder.mapNotNull { it.id }.toSet()
+        val perioderTilSletting = tidligereStønadsperioder.filterNot { gjenståendePerioder.contains(it.id) }
+        stønadsperiodeRepository.deleteAllById(perioderTilSletting.map { it.id })
+    }
+
+    private fun oppdaterStønadsperioder(
+        tidligereStønadsperioder: List<Stønadsperiode>,
+        stønadsperioder: List<StønadsperiodeDto>,
+    ): List<Stønadsperiode> {
+        val stønadsperioderPåId =
+            stønadsperioder.mapNotNull { periode -> periode.id?.let { id -> id to periode } }.toMap()
+        val perioderTilOppdatering = tidligereStønadsperioder.mapNotNull { tidligereStønadsperiode ->
+            stønadsperioderPåId[tidligereStønadsperiode.id]?.let {
+                tidligereStønadsperiode.copy(
                     fom = it.fom,
                     tom = it.tom,
                     målgruppe = it.målgruppe,
                     aktivitet = it.aktivitet,
                 )
-            },
-        ).tilSortertDto()
+            }
+        }
+        return stønadsperiodeRepository.updateAll(perioderTilOppdatering)
+    }
+
+    private fun nyeStønadsperioder(
+        behandlingId: UUID,
+        stønadsperioder: List<StønadsperiodeDto>,
+        tidligereStønadsperiode: List<Stønadsperiode>,
+    ): List<Stønadsperiode> {
+        val tidligereStønadsperiodeIder = tidligereStønadsperiode.map { it.id }.toSet()
+        val nyeStønadsperioder = stønadsperioder.filterNot { tidligereStønadsperiodeIder.contains(it.id) }.map {
+            feilHvis(it.id != null) {
+                "Kan ikke oppdatere stønadsperiode=${it.id} som ikke fra før"
+            }
+            Stønadsperiode(
+                id = UUID.randomUUID(),
+                behandlingId = behandlingId,
+                fom = it.fom,
+                tom = it.tom,
+                målgruppe = it.målgruppe,
+                aktivitet = it.aktivitet,
+            )
+        }
```

Comment:
Usikker på om jeg skjønner hva som skjer her, hva kan trigge denne feilmeldingen? Gir den noe mening for bruker eller er den for oss?

Reply:
`feilHvis` brukes generellt om det er en systemfeil, eller noe vi skal følge med på
Feilmeldingen skjer hvis frontend prøver å sende inn en periode med ID som ikke finnes på behandlingen fra før, finnes test for det

Reply:
Hmm kanskje feilmeldingen heller burde være "Fant ny stønadsperiode med id=id. Nye stønadsperioder skal ikke inneholde id"? Eller noe sånt

Reply:
Det må ikke være fordi man prøver å opprette ny stønadsperiode, men kan også være pga at frontend ikke er oppdatert i en fane, og då prøver å oppdatere noe som ikke lengre eksisterer